### PR TITLE
update provisioner subsystem

### LINKED ISSUE #1143: Guide: Provisioner
Describe a provisioner subsystem for block authorship in the implementor's guide.

We should account for the fact that block authorship as a pipeline is set up as a framework within Substrate, so this subsystem does not actually produce blocks but provides all information to the block producer.

### PR DESCRIPTION
**Pull request by coriolinus**
Closes #1143

### CODE REVIEW

**Comment by coriolinus** [comment]
@rphmeier There are several TODOs in here which I'd appreciate your comments on.

File: roadmap/implementors-guide/src/node/utility/provisioner.md [hunk 1]
```diff
@@ -1,30 +1,42 @@
...
-- dispute inherent
-    > TODO: needs fleshing out in validity module, related to blacklisting
+## Provisionable Data
 
-## Protocol
+### Backable Candidates
+
+The block author can choose 0 or 1 backable parachain candidates per parachain; the only constraint is that each backable candidate has the appropriate relay parent. However, the choice of a backable candidate must be the block author's; the provisioner must ensure that block authors are aware of all available backable candidates.
+
+> TODO: "and their backings": why?
```
Comment (rphmeier) [inline roadmap/implementors-guide/src/node/utility/provisioner.md:13]: the signatures are necessary to submit on-chain. if the `BackedCandidate` type already includes the backing (which I think it does) then disregard

File: roadmap/implementors-guide/src/node/utility/provisioner.md [hunk 2]
```diff
@@ -1,30 +1,42 @@
...
+
+### Signed Bitfields
+
+Signed bitfields are attestations from a particular validator about which candidates it believes are available.
+
+> TODO: Are these actually included in the relay chain block, or just used to help decide whether a block is available and therefore a backable candidate?
```
Comment (rphmeier) [inline roadmap/implementors-guide/src/node/utility/provisioner.md:19]: Yes, included in the relay-chain block. See `Inclusion` and `InclusionInherent` modules

File: roadmap/implementors-guide/src/node/utility/provisioner.md [hunk 3]
```diff
@@ -1,30 +1,42 @@
...
+
+### Misbehavior Reports
 
-Input:
+Misbehavior reports contain proof that a validator or set of validators has misbehaved; they consist of a proof of some kind of misbehavior: double-voting, being the minority vote in a disputed block's vote, etc. These cause dots to be slashed and must be included in the block.
 
-- Bitfield(relay_parent, signed_bitfield)
-- BackableCandidate(relay_parent, candidate_receipt, backing)
-- RequestBlockAuthorshipData(relay_parent, response_channel)
+> TODO: This problem is mentioned elsewhere, but how do we force the block author to include a misbehavior report if they don't like its effects, i.e. they are among the nodes which get slashed?
```
Comment (rphmeier) [inline roadmap/implementors-guide/src/node/utility/provisioner.md:25]: We can't :) but slashing period is long enough that you'll get an honest author soon enough

File: roadmap/implementors-guide/src/node/utility/provisioner.md [hunk 4]
```diff
@@ -1,30 +1,42 @@
...
+
+### Dispute Inherent
+
+> TODO: Is this different from a misbehavior report? How?
```
Comment (rphmeier) [inline roadmap/implementors-guide/src/node/utility/provisioner.md:29]: Unlike a misbehavior report, which is self-contained, the dispute inherent kicks off a dispute process which is resolved over a longer period of time.

**Comment by rphmeier** [comment]
Some more information on handling of the `ProvisionerMessage` would be nice, as well as spin-up and tear-down of jobs, but otherwise LGTM!

File: roadmap/implementors-guide/src/node/utility/provisioner.md [hunk 5]
```diff
@@ -1,30 +1,54 @@
...
+
+Note that there is no mechanism in place which forces a block author to include a misbehavior report which it doesn't like, for example if it would be slashed by such a report. The chain's defense against this is to have a relatively long slash period, such that it's likely to encounter an honest author before the slash period expires.
+
+### Dispute Inherent
+
+The dispute inherent is similar to a misbehavior report in that it is an attestation of misbehavior on the part of a validator or group of validators. Unlike a misbehavior report, it is not self-contained: resolution requires coordinated action by several validators. The canonical example of a dispute inherent involves a fisherman discovering that a set of validators has improperly approved an invalid parachain block: resolving this requires the entire validator set to re-validate the block, so that the minority can be slashed.
+
+> TODO: Is this statement accurate: "Resolution of a dispute produces a misbehavior report."?
```
Comment (rphmeier) [inline roadmap/implementors-guide/src/node/utility/provisioner.md:29]: Somewhat, in the sense that resolution of a dispute will result in the dispute being replayed onto other chains, which can happen in a single transaction. I would expect the provisioner to be responsible for selecting disputes to replay, although that's beyond the scope of this PR.

The validity module as-is elaborates on the distinction between remote and local disputes and the mechanics of replays.

File: roadmap/implementors-guide/src/node/utility/provisioner.md [hunk 6]
```diff
@@ -1,30 +1,54 @@
...
+
+Note that there is no mechanism in place which forces a block author to include a misbehavior report which it doesn't like, for example if it would be slashed by such a report. The chain's defense against this is to have a relatively long slash period, such that it's likely to encounter an honest author before the slash period expires.
+
+### Dispute Inherent
+
+The dispute inherent is similar to a misbehavior report in that it is an attestation of misbehavior on the part of a validator or group of validators. Unlike a misbehavior report, it is not self-contained: resolution requires coordinated action by several validators. The canonical example of a dispute inherent involves a fisherman discovering that a set of validators has improperly approved an invalid parachain block: resolving this requires the entire validator set to re-validate the block, so that the minority can be slashed.
```
Comment (rphmeier) [inline roadmap/implementors-guide/src/node/utility/provisioner.md:27]: I'd broaden this from fisherman to approval checker, which is the language used in the `Validity` module that will handle these disputes.

File: roadmap/implementors-guide/src/type-definitions.md [hunk 7]
```diff
@@ -173,6 +173,34 @@ enum MisbehaviorArbitrationMessage {
...
+## Provisioner Message
+
+```rust
+/// This data becomes intrinsics or extrinsics which should be included in a future relay chain block.
+enum ProvisionableData {
+  /// This bitfield indicates the availability of various candidate blocks.
+  Bitfield(Hash, SignedAvailabilityBitfield),
+  /// The Candidate Backing subsystem believes that this candidate is backable, pending availability.
+  /// TODO: do we need any more data than this?
+  BackableCandidate(Hash, CandidateReceipt),
```
Comment (rphmeier) [inline roadmap/implementors-guide/src/type-definitions.md:185]: Should be the `BackedCandidate` struct, which contains the candidate and its backing. That's what's actually submitted on-chain to the `InclusionInherent` module

**Pull request merged**